File: 10.binary-tree-zigzag-level-order-traversal/binary-tree-zigzag-level-order-traversal.go
package main

import "fmt"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func main() {
	node := initTree()
	ret := zigzagLevelOrder(node)

	for _, arr := range ret {
		for _, value := range arr {
			fmt.Print(value, "\n")
		}
		fmt.Println("===========")
	}
}

func initTree() *TreeNode {
	// 3,9,20,null,null,15,7
	ret := &TreeNode{Val: 3, Left: &TreeNode{Val: 9}, Right: &TreeNode{Val: 20, Left: &TreeNode{Val: 15}, Right: &TreeNode{Val: 7}}}

	return ret
}

func zigzagLevelOrder(root *TreeNode) [][]int {

	res := [][]int{}
	var dfs func(*TreeNode, int)
	dfs = func(root *TreeNode, level int) {
		if root == nil {
			return
		}

		if level >= len(res) {
			res = append(res, []int{})
		}

		if level%2 == 0 {
			// 正向
			res[level] = append(res[level], root.Val)
		} else {
			// 反向
			temp := make([]int, len(res[level])+1)
			temp[0] = root.Val
			copy(temp[1:], res[level])
			res[level] = temp
		}
		dfs(root.Left, level+1)
		dfs(root.Right, level+1)
	}

	dfs(root, 0)
	return res
}

File: 10.binary-tree-zigzag-level-order-traversal/binary-tree-zigzag-level-order-traversal.php
<?php
// ret [[3],[20,9],[15,7]]
function zigzagLevelOrder($arr)
{
    if (empty($arr)) {
        return [];
    }
    return recursive([], [$arr]);
}

function recursive($ret, $arr)
{
    $res = getLevelData($arr);
    if (!empty($res['ret'])) {
        if (count($ret) % 2 == 1) {
            // 翻转
            $temp = [];
            for ($i = count($res['ret']) - 1; $i >= 0; $i--) {
                $temp[] = $res['ret'][$i];
            }
            $res['ret'] = $temp;
        }
        $ret[] = $res['ret'];
    }
    if (!empty($res['nextLevelTree'])) {
        $ret = recursive($ret, $res['nextLevelTree']);
    }
    return $ret;
}

function getLevelData($arr)
{
    $ret           = [];
    $nextLevelTree = [];

    if (!empty($arr)) {
        foreach ($arr as $item) {
            $ret[] = $item['value'];
            if (!empty($item['left'])) {
                $nextLevelTree[] = $item['left'];
            }
            if (!empty($item['right'])) {
                $nextLevelTree[] = $item['right'];
            }
        }
    }
    return array(
        'ret'           => $ret,
        'nextLevelTree' => $nextLevelTree,
    );
}

$arr = [
    'value' => 3,
    'left'  => [
        'value' => 9,
        'left'  => [],
        'right' => [
            'value' => 2,
            'left'  => [
                'value' => 6,
                'left'  => [],
                'right' => [],
            ],
            'right' => [],
        ],
    ],
    'right' => [
        'value' => 20,
        'left'  => [
            'value' => 15,
            'left'  => [],
            'right' => [],
        ],
        'right' => [
            'value' => 7,
            'left'  => [],
            'right' => [],
        ],
    ],
];

var_dump(zigzagLevelOrder($arr));